src/address.c

```diff
@@ -0,0 +1,145 @@
+    int ret;
+
+    if (output)
+        *output = NULL;
+
+    if (!hdkey || (version & ~0xff) || (flags & ~0xff) || !output)
+        return WALLY_EINVAL;
+
+    if (!(flags & (WALLY_ADDRESS_TYPE_P2PKH | WALLY_ADDRESS_TYPE_P2SH_P2WPKH)))
+        return WALLY_EINVAL;
+
+    // Catch known incorrect combinations of address type and version:
+    if ((flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0x05) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x00) ||
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0xC4) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x6F))
+        return WALLY_EINVAL;
+
+    if (flags == WALLY_ADDRESS_TYPE_P2PKH) {
+        // pub_key_hash = ripemd160(sha256(pubkey)
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    } else {
+        // redeem_script = SegWit version 0 + push(keyhash) = OP_0 + 0x20 + [key_hash]
+        // where key_hash = ripemd160(sha256(pubkey))
+        unsigned char redeem_script[HASH160_LEN + 2];
+        redeem_script[0] = OP_0;
+        redeem_script[1] = HASH160_LEN;
+
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), redeem_script + 2, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+
+        // P2SH address = version (e.g. 0x05) + ripemd160(sha256(redeem_script))
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(redeem_script, sizeof(redeem_script), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    }
+
+    ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+
+    wally_clear(address, sizeof(address));
+    return ret;
+}
+
+int wally_bip32_key_to_addr_segwit(const struct ext_key *hdkey, const char *addr_family,
+                                   uint32_t flags, char **output)
+{
+    int ret;
+
+    // Witness program bytes, including the version and data push opcode.
+    unsigned char witness_program_bytes[HASH160_LEN + 2];
+    witness_program_bytes[0] = OP_0;
+    witness_program_bytes[1] = HASH160_LEN;
+
+    if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), witness_program_bytes + 2, HASH160_LEN) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    ret = wally_addr_segwit_from_bytes(witness_program_bytes, HASH160_LEN + 2, addr_family, flags, output);
+    return ret;
+}
+
+int wally_address_to_scriptpubkey(const char *addr, uint32_t flags, unsigned char *bytes_out,
+                                  size_t len, size_t *written)
+{
+    int ret;
+
+    uint32_t version;
+    unsigned char bytes_base58_decode[len + 1];
+    size_t written_base58_decode;
+
+    if (written)
+        *written = 0;
+
+    if ((flags & ~0xff))
+        return WALLY_EINVAL;
+
+    if (!(flags & (WALLY_NETWORK_BITCOIN_MAINNET | WALLY_NETWORK_BITCOIN_TESTNET)))
+        return WALLY_EINVAL;
```

Comment:
It may be simpler to just check that "flags" is equal to either WALLY_NETWORK_BITCOIN_MAINNET or WALLY_NETWORK_BITCOIN_TESTNET and then drop the check on line 87/88 as well.

Reply:
Done